src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
+  | ExpectedEndOfFile i
+  | CustomError e
+  | Empty
+  deriving (Eq)
+
+instance (Show i, Show e) => Show (ParseErrorType i e) where
+  show :: ParseErrorType i e -> String
+  show EndOfInput = "End of Input"
+  show (Unexpected a) = "Unexpected " ++ show a
+  show (Expected e a) = "Expected " ++ show e ++ ", got " ++ show a
+  show (ExpectedEndOfFile a) = "Expected EndOfFile, got " ++ show a
+  show (CustomError e) = "Custom Error: " ++ show e
+  show Empty = "No error"
+
+--------------------
+---- ParseError ----
+--------------------
+
+type Offset = Integer
+
+data ParseError i e where
+  ParseError
+    :: { errOffset :: Offset
+       , errError :: ParseErrorType i e
+       }
+    -> ParseError i e
+
+
+instance (Show i, Show e) => Show (ParseError i e) where
+  show :: ParseError i e -> String
+  show (ParseError o e) = "Error: " ++ show e ++ ", at offset " ++ show o
+
+----------------
+---- Parser ----
+----------------
+
+newtype Parser i m e a where
+  Parser
+    :: { parse
+          :: [i]
+          -> Offset
+          -> Either
+              (m (ParseError i e))
```

Comment:
As we previoulsy discussed, I'm not sure how valuable having a type variable `m` for the container that accumulates the errors is. Did you get any ideas to use something other than `[]`?

Reply:
I was thinking of a making a more sophisticated type that accumulates errors more intelligently